src/bridges/aave/lending/AaveLendingBridge.sol

```diff
@@ -0,0 +1,342 @@
+         * 3. Deposit assets into Aave (receives aUnderlyingAsset in return)
+         * 4. Mint zkATokens equal to scaled amount
+         * 5. Approve ROLLUP_PROCESSOR to pull funds
+         */
+
+        if (isEth) {
+            WETH.deposit{value: amount}();
+        }
+
+        ILendingPool pool = ILendingPool(ADDRESSES_PROVIDER.getLendingPool());
+        // Compute scaledAmount rounded down. Aave uses rayDiv which can round up and down (at 0.5).
+        // For consistency we round down.
+        uint256 scaledAmount = (amount * 1e27) / pool.getReserveNormalizedIncome(underlyingAsset);
+        require(scaledAmount > 0, Errors.ZERO_VALUE);
+
+        IERC20(underlyingAsset).safeApprove(address(pool), amount);
+        pool.deposit(underlyingAsset, amount, address(this), 0);
+
+        IAccountingToken zkAToken = IAccountingToken(zkATokenAddress);
+        zkAToken.mint(address(this), scaledAmount);
+        zkAToken.approve(ROLLUP_PROCESSOR, scaledAmount);
+
+        return scaledAmount;
+    }
+
+    /**
+     * @notice Withdraw `underlyingAsset` from Aave
+     * @param underlyingAsset The address of the underlying asset
+     * @param scaledAmount The amount of zkAToken to burn, used to derive underlying amount
+     * @return The underlying amount of tokens withdrawn
+     */
+    function _exit(
+        address underlyingAsset,
+        address zkATokenAddress,
+        uint256 scaledAmount,
+        uint256 interactionNonce,
+        bool isEth
+    ) internal returns (uint256) {
+        /**
+         * Interaction flow:
+         * 0. Burn zkATokens equal to scaledAmount
+         * 1. Compute the amount of underlying assets to withdraw
+         * 2. Withdraw from the Aave pool
+         * 3. If underlyingAsset is supposed to be ETH, unwrap WETH
+         * 4. Approve underlying asset to be pulled by ROLLUP_PROCESSOR or transfer ETH
+         * Exit may fail if insufficient liquidity is available in the Aave pool.
+         */
+
+        IAccountingToken(zkATokenAddress).burn(scaledAmount);
+
+        ILendingPool pool = ILendingPool(ADDRESSES_PROVIDER.getLendingPool());
+        // Compute the underlying amount rounded down. Aave uses rayMul which can round up and down (at 0.5)
+        // For consistency we round down. Will leave aToken dust.
+        uint256 underlyingAmount = (scaledAmount * pool.getReserveNormalizedIncome(underlyingAsset)) / 1e27;
+        require(scaledAmount > 0, Errors.ZERO_VALUE);
+
+        /// Return value by pool::withdraw() equal to underlyingAmount, unless underlying amount == type(uint256).max;
+        uint256 outputValue = pool.withdraw(underlyingAsset, underlyingAmount, address(this));
+
+        if (isEth) {
+            WETH.withdraw(outputValue);
+            IRollupProcessor(ROLLUP_PROCESSOR).receiveEthFromBridge{value: outputValue}(interactionNonce);
+        } else {
+            IERC20(underlyingAsset).safeApprove(ROLLUP_PROCESSOR, outputValue);
```

Comment:
I would approve the token when registered to save gas.

Reply:
Approvals moved to a separate function to ensure that we can re-approve in case Aave updates their proxy address for some reason. Approvals performed when listing asset initially as well. See https://github.com/AztecProtocol/aztec-connect-bridges/pull/21/commits/7053a154f0360e3eadd3ed3a1f9403fff90854f9